### PR TITLE
Doc update to include type evolution rules

### PR DESCRIPTION
**Pull request by chadwalters**
Type evolution rules added to C++ and C# manuals. Duplication will be addressed when https://github.com/Microsoft/bond/issues/115 is addressed.

### CODE REVIEW

File: doc/src/bond_cpp.md [hunk 1]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
 
 See example: `examples/cpp/core/marshaling`.
 
+Type evolution
+==============
+
+Bond does not use explicit versioning to deal with changes to schemas over time
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: Minor: missing a period

File: doc/src/bond_cpp.md [hunk 2]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+- Changing type between `int32` and `enum`
+- Changing type between `vector<T>` and `list<T>`
+- Changing type between `blob` and `vector<uint8>` or `blob` and `list<uint8>`
+- Changing type between `T` and `bonded<T>`
+- Adding new enumeration values at the end of an enumeration
+
+The following changes to a type are generally safe but require some
+consideration about how the change is rolled out:
+
+- Changing between `optional` and `required_optional` or between
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: Minor: multi-line list items are floated in this document:

```
- This is the first line of the list item. It
  keeps going on the second line but is
  indented.
- Here's the second item.
```

File: doc/src/bond_cpp.md (hunk 1 above)
Comment (chwarr) [inline doc/src/bond_cpp.md]: "changed to a type's schema"? We're being inconsistent about using the word "type" and "schema".

File: doc/src/bond_cpp.md [hunk 3]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
 
 See example: `examples/cpp/core/marshaling`.
 
+Type evolution
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: It won't let me add a comment where this should be, but at the top, consider linking the phrase "schema versioning rules" to here.

File: doc/src/bond_cpp.md [hunk 4]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+==============
+
+Bond does not use explicit versioning to deal with changes to schemas over time
+Instead, Bond supports certain type evolution operations which allow the
+producer and consumer of Bond types to evolve without lockstep coordination.
+
+The following changes to a type will never break compatibility across the wire:
+
+- Adding or removing an `optional` or `optional_required` field
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: `required_optional` is the keyword.

File: doc/src/bond_cpp.md [hunk 5]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+- Changing type between `int32` and `enum`
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: "Changing a field's type" here and below?

File: doc/src/bond_cpp.md [hunk 6]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+- Changing type between `vector<T>` and `list<T>`
+- Changing type between `blob` and `vector<uint8>` or `blob` and `list<uint8>`
+- Changing type between `T` and `bonded<T>`
+- Adding new enumeration values at the end of an enumeration
+
+The following changes to a type are generally safe but require some
+consideration about how the change is rolled out:
+
+- Changing between `optional` and `required_optional` or between
+`required_optional` and `required`. This facilitates a two-step process for
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: Consider "The `required_optional` modifier facilitates..."

File: doc/src/bond_cpp.md [hunk 7]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+changing between `optional` and `required` but the first step must be completed
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: "but, the first step" (add comma)

File: doc/src/bond_cpp.md [hunk 8]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+on both the producer and the consumer sides before the second step can be
+started on either side.
+- Promoting a numeric type from a smaller size to a larger size (e.g.: `float`
+to `double`, `uint8` to `uint16`, `uint8` to `uint32`, `int8` to `int16`,
+etc.). The consumer must get the change before the producer.
+- Promoting from `int8` or `int16` to `enum`. The consumer must get the change
+before the producer.
+
+These following changes will break wire compatibility and are not recommended:
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: "The following"?

File: doc/src/bond_cpp.md [hunk 9]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+
+- Adding or removing `required` fields
+- Incompatible change of field types (any type change NOT covered above): e.g.:
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: Odd punctuation at the end of the line.

File: doc/src/bond_cpp.md [hunk 10]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+`int32` to `string`, `string` to `wstring`
+- Changing of field ordinals
+- Changing of inheritance hierarchy (add/remove/change base struct)
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: I would just say "add/remove base struct". You can change the fields inside a base struct if you follow the rules above.

File: doc/src/bond_cpp.md [hunk 11]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+- Changing between `required` and `optional` directly
+- Changing the default value of a field
+- Changing the values of an enumeration in any way other than adding additional
+values at the end
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: Might want to mention that this is because renumbering might happen.

File: doc/src/bond_cpp.md [hunk 12]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+producer and consumer of Bond types to evolve without lockstep coordination.
+
+The following changes to a type will never break compatibility across the wire:
+
+- Adding or removing an `optional` or `optional_required` field
+- Changing type between `int32` and `enum`
+- Changing type between `vector<T>` and `list<T>`
+- Changing type between `blob` and `vector<uint8>` or `blob` and `list<uint8>`
+- Changing type between `T` and `bonded<T>`
+- Adding new enumeration values at the end of an enumeration
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: You _can_ add an enumeration value in the middle, so long as it doesn't change the value of anything else. This is safe:

    enum Foo { Bar = 3; Baz = 2; }
    enum Foo { Bar = 3; Qux = 7; Baz = 2; }

"Adding new enumeration contants without perturbing the numeric value of the existing constants."?

We use "constant" to refer to the things inside the enum in the compiler documentation. Though, I often refer to them as enum member of enumerated values. Alas, enumerant isn't a word.

File: doc/src/bond_cpp.md [hunk 13]
```diff
@@ -185,6 +185,59 @@ particular protocol reader:
...
+- Adding or removing `required` fields
+- Incompatible change of field types (any type change NOT covered above): e.g.:
+`int32` to `string`, `string` to `wstring`
+- Changing of field ordinals
+- Changing of inheritance hierarchy (add/remove/change base struct)
+- Changing between `required` and `optional` directly
+- Changing the default value of a field
+- Changing the values of an enumeration in any way other than adding additional
+values at the end
+
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: Can I change between unsigned to signed if the width is OK? E.g., uint8 to int16 can be done without loss.

File: doc/src/bond_cs.md [hunk 14]
```diff
@@ -261,6 +261,59 @@ See also the following example:
 
 - `examples/cs/core/marshaling`
 
+Type evolution
```
Comment (chwarr) [inline doc/src/bond_cs.md]: I didn't look at this, assuming it's the same as the C++.

File: doc/src/bond_cpp.md (hunk 3 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: Done

File: doc/src/bond_cpp.md (hunk 1 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: Done

Comment (chadwalters) [inline doc/src/bond_cpp.md]: I've worked on this a bit.

File: doc/src/bond_cpp.md (hunk 4 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: Done.

File: doc/src/bond_cpp.md (hunk 12 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: I will switch to use "constants" and clarify allowed operations.

File: doc/src/bond_cpp.md (hunk 6 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: Done

File: doc/src/bond_cpp.md (hunk 7 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: That seems very odd stylistically to me -- comma before the 'but' perhaps but not after. See point 2 of http://grammar.ccc.commnet.edu/Grammar/commas.htm

File: doc/src/bond_cpp.md (hunk 10 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: "change" -> "substitute"? Someone might think they can change parent from one struct to another ("hey I didn't add or remove anything").

File: doc/src/bond_cpp.md (hunk 11 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: Reworded to allow for possibility you mentioned above where the change doesn't need to come at the end.

File: doc/src/bond_cpp.md (hunk 13 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: That is a good question. I need to test that.

File: doc/src/bond_cs.md (hunk 14 above)
Comment (chadwalters) [inline doc/src/bond_cs.md]: Yes.

File: doc/src/bond_cpp.md (hunk 7 above)
Comment (chwarr) [inline doc/src/bond_cpp.md]: Yes, I meant ", but the first step". Good catch.

File: doc/src/bond_cpp.md [hunk 15]
```diff
@@ -185,6 +185,65 @@ particular protocol reader:
...
+  `list<uint8>`
+- Changing a field's type between `T` and `bonded<T>`
+- Adding new enumeration constants that don't alter existing constants (beware
+  of implicit reordering)
+
+The following changes to a type are generally safe but require some
+consideration about how the change is rolled out:
+
+- Changing a field between `optional` and `required_optional` or between
+  `required_optional` and `required`. The `required_optional`. modifier
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: Minor: it looks like there's an extraneous period after second `required_optional`.

File: doc/src/bond_cpp.md [hunk 16]
```diff
@@ -185,6 +185,65 @@ particular protocol reader:
...
+  facilitates a two-step process for changing between `optional` and
+  `required`, but the first step must be completed on both the producer and the
+  consumer sides before the second step can be started on either side.
+- Promoting a field with a numeric type from a smaller size to a larger size
+  (e.g.: `float` to `double`, `uint8` to `uint16`, `uint8` to `uint32`, `int8`
+  to `int16`, etc.). The consumer must get the change before the producer.
+  Changing from unsigned to signed or vice versa is NOT compatible (e.g.:
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: Since the changing the signedness will break, should that prohibition be below?

Comment (chwarr) [inline doc/src/bond_cpp.md]: Minor: `*not*` will render nicer than `NOT`.

* `*not*` *not*
* `NOT` NOT

File: doc/src/bond_cpp.md [hunk 17]
```diff
@@ -185,6 +185,65 @@ particular protocol reader:
...
+  `uint8` to `int16`).
+- Promoting from `int8` or `int16` to `enum`. The consumer must get the change
+  before the producer.
+
+These following changes will break wire compatibility and are not recommended:
+
+- Adding or removing `required` fields
+- Incompatible change of field types (any type change NOT covered above); e.g.:
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: Minor: same comment about `NOT` here.

File: doc/src/bond_cpp.md [hunk 18]
```diff
@@ -185,6 +185,65 @@ particular protocol reader:
...
+- Changing between `required` and `optional` directly
+- Changing the default value of a field
+- Changing existing enumeration constants in any way (including implicit
+  renumbering)
+
+Some best practices and other considerations to keep in mind:
+
+- When removing a field, comment it out rather than removing it altogether so
+that the field ordinal is not reused in future edits of the schema
+- When working with untagged protocols like SimpleBinaryProtocol, great care
```
Comment (chwarr) [inline doc/src/bond_cpp.md]: Might want to mention that runtime SchemaDefs can be used with a link to the SimpleBinaryProtocol section.

File: doc/src/bond_cs.md [hunk 19]
```diff
@@ -261,6 +261,64 @@ See also the following example:
 
 - `examples/cs/core/marshaling`
 
+Schema evolution
```
Comment (chwarr) [inline doc/src/bond_cs.md]: Again, didn't look at this, assuming that it's the same as the C++ one.

File: doc/src/bond_cpp.md (hunk 15 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: Done.

File: doc/src/bond_cpp.md (hunk 16 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: No, I think it belongs here as a caveat on the compatible ones.

Comment (chadwalters) [inline doc/src/bond_cpp.md]: Done

File: doc/src/bond_cpp.md (hunk 17 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: Done

File: doc/src/bond_cpp.md (hunk 18 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: I linked to the runtime schema section.

File: doc/src/bond_cs.md (hunk 19 above)
Comment (chadwalters) [inline doc/src/bond_cs.md]: Yes

File: doc/src/bond_cpp.md (hunk 18 above)
Comment (chadwalters) [inline doc/src/bond_cpp.md]: And the SimpleBinary section

**Comment by chadwalters** [comment]
I added a protocols section to the C# manual. This material was largely missing on the C# side.

**Comment by chadwalters** [comment]
CI test failures not related to change.
Please "Rebase and merge" (don't squash).

**Pull request merged**